### ISSUE TITLE
running PECAN

### ISSUE THREAD

**Issue by fslee62** [vamships/PECAN#4]
hello srivamshi,
i was learning about GCN and came across your paper. i find it to be highly relevant to what i am interested in. so i decided to try out PECAN. i am a beginner in NN. so my questions may reflect a lack of understanding in general. here are a few initial questions:
1. as far as i can tell, the dataset (the 191 pro-pro complexes) for transfer learning is not included in the download from zenodo.org. did i miss something here or it is available somewhere?
2. the matrix dimensions inside the provided pickle files (e.g. train.cpkl) are different from what was described in the README file. for example, l_edge is 'numresx25' (cpkl) and 'numresx25x2' (readme). also, l_hood_indices is 'numresx30' (cpkl) and 'numresx25' (readme). did i misinterpret something?
3. my understanding is that i need to use the pro-pro dataset to run GCN_protein_base. only after that can i then run GCN_xTransfer, which uses output from GCN_protein_base. am i correct here? if so, don't i need to have access to the 191 pro-pro dataset? currently, the yml files inside GCN_protein_base use "train.cpkl" as input, which i don't understand.
4. thinking ahead, i like to be able to create my own pickle input files, which looks like a daunting task to me. is it possible that you provide more information on how to achieve that like a procedure and some auxiliary tools/programs. if not, any pointers here.

i know that is a handful. i sincerely appreciate your help and guidance. cheers.

**Comment by vamships** [comment]
Hi @fslee62,

Thanks for your interested in this work.

1. Yes, the data for training the protein-protein model is not included here. Since the entire DBD v5 was available, we just included the data for epitope and paratope prediction. All one would need to do is discard the antibody and antigen complexes to ensure fairness.
2. The dimensions were correct when I last checked. The contents of the variables `l_edge` and `r_edge` in my files are empty (zeros), so you could create your own matrices of the correct dimensions externally. `l_hood_indices` and `r_hood_indices` contain the indices of spatial neighbors. They could have been padded with dummy residue indices. You could ignore those beyond 25.
3. Yes, you'd need a file `train.cpkl` that contains protein-protein complexes to train a base model, which could then be used for transfer learning. You could create one from DBD v5 or any database that provides such complexes.
4. The [pickle library](https://docs.python.org/3/library/pickle.html#data-stream-format) in python is pretty straightforward to use. The functions `dump()` and `load()` are probably the only two you'd use to write and read data respectively.  A tutorial could be found [here](https://youtu.be/XzkhtWYYojg) and [here](https://realpython.com/python-pickle-module/).

Hope this helps.

**Comment by fslee62** [comment]
hi srivamshi,
thanks for the quick responses. they clarified things for me. 

with respect to the pickle input, i was referring to the content of the file like the content of l_vertex for instance. yes reading and writing a pickle file is not a concern to me. is there anything special or particular about those calculations; e.g., calculating PSSM, solvent accessibility, etc. i would guess that those numbers would be sensitive to the parameters and methods used.

indeed, to create the pro-pro pickle file, i would need to calculate all those quantities, which was the part that seemed daunting to me. i do have the whole DBD5.5 dataset. would you have a write-up on how to calculate those quantities for the various matrices?

i do have a couple of new questions on how to use/read the results. say i want to know who (which residues) are the interface residues in each case.

(1) is "output_preds.csv" inside each of those x.ann directories under test_4/ the most important file to look at? or i should look at other files.
(2) do i convert those numbers inside output_preds.csv into 1 or -1? i saw the get_predicted_label() in your answer to another issue. is that function what i use to map the output_preds.csv numbers into 1's and -1's?

thanks again for your help.
fred

**Comment by fslee62** [comment]
i managed to resolve the questions listed above and was able to reproduce the AUC-PR numbers for the Task-specific learning (GCN_task) cases given in Table2 and Table3 of the bioRxiv preprint (https://doi.org/10.1101/658054). therefore, i am closing this issue and thank the author for all the help.

**Comment by vamships** [comment]
Hi @fslee62 , 
Sorry for the delay in getting back to you. I am glad that you were able to reproduce the results from the paper. To generate the features similar to those in the pickle files, you could follow the *Input Representation* subsection of the paper. Unfortunately, there's no single script that generates those features from PDB files.

(1) If you are interested in node classification as shown in the paper, yes. It represents per-residue score of being an interface.
(2) To assign a class label, you'd need to choose a cutoff on the predicted scores. Generally, 0.5 is the cutoff, but it can be adjusted to other values to trade-off between precision and recall. Setting the cutoff by protein sizes could also be a good strategy.

**Comment by fslee62** [comment]
setting cutoffs based on protein sizes is an interesting idea. will look into that. cheers.
